drbg/src/hashdrbg.rs

```diff
@@ -0,0 +1,220 @@
+    }
+    /// This is the `Hashgen()` function from _10.1.1.4_.
+    fn hash_gen(&self, out: &mut [u8]) {
+        let m = out.len() / Hash::OutputSize::USIZE;
+        let p = U896::from_u64(2).shl(Hash::SeedLenBytes::USIZE * 8);
+
+        let mut data = self.v.clone();
+        for i in 0..m {
+            let w = Self::hash([&data].iter());
+            out[i * Hash::OutputSize::USIZE..(i + 1) * Hash::OutputSize::USIZE].copy_from_slice(&w);
+            let mut x = slice_to_u896(&data);
+            x = x.add_mod(&U896::ONE, &p);
+            data = truncate_from_start(&x.to_be_byte_array());
+        }
+        let remainder = out.len() % Hash::OutputSize::USIZE;
+        // we may need to do the hashing on last time
+        if remainder != 0 {
+            let w = Self::hash([&data].iter());
+            out[m * Hash::OutputSize::USIZE..].copy_from_slice(&w[..remainder]);
+        }
+    }
+}
+
+impl<Hash: Digest + HashDrbgProps> Drbg for HashDrbg<Hash> {
+    type Entropy = GenericArray<u8, Hash::EntropyLenBytes>;
+    type Nonce = GenericArray<u8, Hash::NonceLenBytes>;
+    /// according to Table 2: Definitions for Hash-Based DRBG Mechanisms,
+    /// instantiate should fail if `personalization_string` contains more
+    /// than `2**35` bits which is ~ 4GB.
+    #[allow(refining_impl_trait)]
+    fn instantiate(
+        entropy_input: &Self::Entropy,
+        nonce: &Self::Nonce,
+        personalization_string: &[u8],
+    ) -> Result<HashDrbg<Hash>, DrbgError> {
+        if personalization_string.len() > MAX_BYTES_PERSONALIZATION_STRING {
+            return Err(DrbgError);
+        }
+        Ok(Self::instantiate_inner(
+            entropy_input,
+            nonce,
+            personalization_string,
+        ))
+    }
+    /// according to Table 2: Definitions for Hash-Based DRBG Mechanisms,
+    /// instantiate should fail if `additional_input` contains more
+    /// than `2**35` bits which is ~ 4GB.
+    fn reseed(
+        &mut self,
+        entropy_input: &Self::Entropy,
+        additional_input: &[u8],
+    ) -> Result<(), DrbgError> {
+        if additional_input.len() > MAX_BYTES_ADDITIONAL_INPUT {
+            return Err(DrbgError);
+        }
+        self.reseed_inner(entropy_input, additional_input);
+        Ok(())
+    }
+    /// according to Table 2: Definitions for Hash-Based DRBG Mechanisms,
+    /// instantiate should fail if `additional_input` contains more
+    /// than `2**35` bits which is ~ 4GB.
+    fn next_u32(&mut self, additional_input: &[u8]) -> Result<u32, DrbgError> {
+        next_u32_via_fill(self, additional_input)
+    }
+    /// according to Table 2: Definitions for Hash-Based DRBG Mechanisms,
+    /// instantiate should fail if `additional_input` contains more
+    /// than `2**35` bits which is ~ 4GB, we ignore that check since it is
+    /// not possible to pass that many data into a tpm chip.
+    fn next_u64(&mut self, additional_input: &[u8]) -> Result<u64, DrbgError> {
```

Comment:
Duplicate comment or missing check on additional_input size?

Reply:
check is done implicitly by the the `next_u64_via_fill`